add phase 1 report of tanmoysrt (#166)
CircuitVerse/Blog: content/posts/Tanmoy_Sarkar_Phase_1_Report.md

---
title: "Improve Development Experience: Phase 1 Report (GSoC 2023)"
date: 2023-07-07T12:04:13+05:30
draft: false
author: Tanmoy Sarkar
type: post
---

> This blog is dedicated to my progress in the first phase of Google Summer of Code  2023.

Reviewer: ```suggestion
This blog is dedicated to summarise the progress for the first phase of the project "Improve the Development Experience" in Google Summer of Code'2023.
```

## About My Project

> My project focuses on making the process of working with CircuitVerse easier and more efficient. The main goal is to simplify setting up the development environment and working with the code. Additionally, Our aim to enhance the test cases and improve the continuous integration (CI) workflow, which will ultimately reduce the time required for code review by maintainers.

Reviewer: ```suggestion
This project focuses on making the process of working and developing with CircuitVerse easier and more efficient. The main goal is to simplify setting up the development environment and working with the code. Additionally, Our aim to enhance the test cases and improve the continuous integration (CI) workflow, which will ultimately reduce the time required for code review by maintainers.
```

In the first phase, we have covered

* Integrate Ruby Debugger [Week 1]
* Integrate Solargraph LSP [Week 2]
* Integrate Vite Rails [Week 3]
* RBS Installation & CI Integration [Week 3]
* Integrate Undercover CI [Week 4]
* Improve Unit testcase coverage [Week 5]
* Improve integration testing [Week 6]

### Integrate Ruby Debugger

> This integration enhances the debugging experience for developers by providing a visual debugging capability directly from their integrated development environment (IDE). It allows anyone to conveniently set breakpoints and debug the application from their IDE. This integration support  also docker based environment. By changing configuration in IDE , it can easily connect to TCP based debugger and debug the application easily. Documentation has been updated accordingly.

Reviewer: ```suggestion
This integration enhances the debugging experience for developers by providing a visual debugging capability directly from their integrated development environment (IDE). It allows anyone to conveniently set breakpoints and debug the application from their IDE. This integration also support docker based environment. By changing configuration in IDE , it can easily connect to TCP based debugger and debug the application easily. The documentation has been updated accordingly.
```

Could you also embed `Document` with respective doc link

Author: as the PR has not yet merged, should I embed the PR branch's documentation?

Reviewer: No worries, we can add that after!

This integration also supports Docker-based environments, allowing the application to be run in a Docker container while connecting the IDE/Editor's solargraph extension to the Solargraph server via a TCP socket. Initially, there was a significant challenge in this setup. The extension used to send the current directory to the LSP server, but the LSP server had a different path inside the Docker container. However, this issue was successfully resolved by creating a symlink inside the container. 

![solargraph-docker](/images/tanmoy_gsoc23/solargraph-docker.jpeg)

For more detailed information, please refer to the blog post available at [https://tanmoy.online/week-1-2-circuitversegsoc23](https://tanmoy.online/week-1-2-circuitversegsoc23)

Currently, documentation and support has been added for 

* VS Code
* Sublime Editor
* Neovim

##### References
- Demo Video : [Click Here](https://github.com/CircuitVerse/CircuitVerse/assets/57363826/5ec17053-e7fe-4bfd-abb3-673b54711de1)
- All PR's : [Integrate LSP with Solargraph](https://github.com/CircuitVerse/CircuitVerse/issues/3761)

### Integrate Vite Rails

> Previously in the project all the javascript compiled in single one via esbuild and served by sprocket asset pipeline in development and production [mostly same]. As a result, during development on each change, esbuild need to build multiple JS files againg which increase friction during development. 

Reviewer: ```suggestion
Previously in the project, all javascript compiled in a single one via esbuild and served by sprocket asset pipeline in development and production [mostly same]. As a result, during development on each change, esbuild need to build multiple JS files againg which increase friction during development. 
```
- @tanmoysrt is `againg` a typo?

> That’s why vite rails integrated and simulator assets moved to vite. After this integration, if there are many changes in the javascript, only the spciefic portion will be updated i website and hot reload will be performed if possible, else complete page will reload. Also, instead of packing and serving javascript files, it will load all dependency of javascript files asynchronously in browser. 

Reviewer: ```suggestion
That’s why vite rails is integrated and simulator assets moved to vite. After this integration, if there are many changes in the javascript, only the specific portion will be updated in the website and hot reload will be performed (if possible), otherwise entire page will reload. Also, instead of packing and serving javascript files, it will load all dependency of javascript files asynchronously in browser. 
```

In production, vite can pack all the javascript files together and serve static javascript files to make it faster.

##### References
- Demo Video : [Click Here](https://github.com/CircuitVerse/CircuitVerse/assets/57363826/e1836945-5aff-4c13-b4ac-375483291bec)
- PR : [Move Simulator Assets To Vite Rails](https://github.com/CircuitVerse/CircuitVerse/pull/3777)

### RBS Installation & CI Integration

RBS [Ruby static typing] enabled static typing mechanism which enables to maintain consistency in codebase and reduce the bugs in code. 

> We used `rbs_rails` gem for rbs as it provides some prebuilt task and cli to generate untyped *.rbs annotation files from already written codes. With this we used `steep` gem to verify type signatures. 

Reviewer: ```suggestion
We used `rbs_rails` gem for rbs as it provides some prebuilt task and cli to generate untyped `*.rbs` annotation files from already written codes. With this we used `steep` gem to verify type signatures. 
```

- Demo Video : [Click Here](https://github.com/CircuitVerse/CircuitVerse/assets/57363826/e1836945-5aff-4c13-b4ac-375483291bec)
- PR : [Move Simulator Assets To Vite Rails](https://github.com/CircuitVerse/CircuitVerse/pull/3777)
- PR : [Install RBS & Configure](https://github.com/CircuitVerse/CircuitVerse/pull/3807)
- PR : [feat: Add RBS to CI](https://github.com/CircuitVerse/CircuitVerse/pull/3833)
- PR : [Integrate Undercover CI](https://github.com/CircuitVerse/CircuitVerse/pull/3812)
- PR : [Unit test for models](https://github.com/CircuitVerse/CircuitVerse/pull/3835)
- PR : [Unit test for helpers](https://github.com/CircuitVerse/CircuitVerse/pull/3836)
- PR : [Unit test for controllers](https://github.com/CircuitVerse/CircuitVerse/pull/3852)
- PR : [Unit test for jobs](https://github.com/CircuitVerse/CircuitVerse/pull/3853)
- PR : [Integration testing for profile management](https://github.com/CircuitVerse/CircuitVerse/pull/3863)
- PR : [Integration testing for group management](https://github.com/CircuitVerse/CircuitVerse/pull/3864)
- PR : [Integration testing for assignment management](https://github.com/CircuitVerse/CircuitVerse/pull/3865)

> - PR : [Integration testing for project management](https://github.com/CircuitVerse/CircuitVerse/pull/3866)

Reviewer: Could you also include a small glimpse of next phase and a conclusion.

In the first phase, we have covered

* Integrate Ruby Debugger [Week 1]
* Integrate Solargraph LSP [Week 2]
* Integrate Vite Rails [Week 3]
* RBS Installation & CI Integration [Week 3]
* Integrate Undercover CI [Week 4]
* Improve Unit testcase coverage [Week 5]
* Improve integration testing [Week 6]

### Integrate Ruby Debugger

This integration enhances the debugging experience for developers by providing a visual debugging capability directly from their integrated development environment (IDE). It allows anyone to conveniently set breakpoints and debug the application from their IDE. This integration also support docker based environment. By changing configuration in IDE , it can easily connect to TCP based debugger and debug the application easily. [The documentation](https://github.com/tanmoysrt/CircuitVerse/blob/solargraph/yard/documentation/LSP-SETUP.MD) has been updated accordingly.

##### References

> - Demo Video : [Click Here](https://github.com/CircuitVerse/CircuitVerse/assets/57363826/1c0e957e-3eed-4555-8670-27df282786ab)

Reviewer 2: Add all videos and images directly on this repo as the links might break in future

This integration enhances the debugging experience for developers by providing a visual debugging capability directly from their integrated development environment (IDE). It allows anyone to conveniently set breakpoints and debug the application from their IDE. This integration also support docker based environment. By changing configuration in IDE , it can easily connect to TCP based debugger and debug the application easily. [The documentation](https://github.com/tanmoysrt/CircuitVerse/blob/solargraph/yard/documentation/LSP-SETUP.MD) has been updated accordingly.

**Debugger Preview**

> ![ruby-debugger](/images/tanmoy_gsoc23/ruby_debugger.gif)

Reviewer 2: Could you make this a video instead of gif, this seems to be 10+MB, automatically loading this gif on the page will slow the page load

Author: Should I add link of video only , as in markdown video support is not there ?

Reviewer 3: There are 2 ways you can do this

1. Create your own shortcode named `video.html` in `layouts/shortcodes/video.html`

```html
{{ $videoPath := index .Params 0 }}

<div style="position: relative; padding-bottom: 56.25%; height: 0; overflow: hidden; max-width: 100%; margin: 0 auto;">
  <iframe src="{{ $videoPath }}" width="100%" height="100%" frameborder="0" allowfullscreen style="position: absolute; top: 0; left: 0;"></iframe>
</div>
```
Or use `video` tag instead of `iframe` and get creative with the CSS.

And call it in your markdown as

```bash
{{< video "/images/vaibhav-upreti/ruby_debugger.mp4" >}}
```

2. Use in built hugo `youtube` shortcode that this theme provides by default 

For this youtube video
https://www.youtube.com/watch?v=VmpgOhdMA2g

we can embed it as.... 
```bash
{{< youtube VmpgOhdMA2g >}}
```

Author: Thanks @vaibhavupreti

Author: @tachyons updated the blog !
